File: app/GOExercise/GOExercise/Services/StreakFreezeWindow.swift
import Foundation

/// 連続が途切れた後「4日グレース」内で、フリーズ(rescue ticket)を missed 日に
/// 適用すれば連続が復活する状況を検出する純ロジック。
/// 2層: `Decision`(status 列だけの純判定・swiftc 完全テスト可)+ records 入口。
enum StreakFreezeWindow {
    struct Result: Equatable {
        /// 復活可能(直近グレース内に missed 日があり、その手前に連続の頭がある)。
        let revivable: Bool
        /// フリーズを当てるべき「今日からの日数オフセット」(1=昨日, 2=一昨日…)。
        let missedOffsets: [Int]
        /// 必要フリーズ枚数(= missed 日数)。
        let freezesNeeded: Int
        /// 残枠が足りるか。
        let hasEnough: Bool
    }

    enum Decision {
        /// statuses[0]=昨日, [1]=一昨日… の順(古い方向)。
        /// 直近 `lookback` 日ぶんの missed を集めつつ、**その手前(最大 lookback+1 日目)に連続の頭**
        /// (achieved/rest 経由)があるかで復活可否を決める。lookback 丁度(例: 4日)の連続欠けでも、
        /// 1つ先(offset lookback+1)の achieved を anchor に拾えるよう statuses は lookback+1 件渡す。
        static func evaluate(statuses: [DailyStatus], remainingFreezes: Int, lookback: Int) -> Result {
            var missedOffsets: [Int] = []
            var foundPrior = false
            var tooOld = false // missed が lookback を超えて続く = 4日グレース外で復活不可
            loop: for (idx, status) in statuses.enumerated() {
                let offset = idx + 1
                switch status {
                case .achieved, .todayAchieved:
                    foundPrior = true // 連続の頭に到達
                    break loop
                case .rest:
                    continue // freeze 不要、連続も切らない → 次へ
                case .missed:
                    if offset <= lookback {
                        missedOffsets.append(offset) // offset 1=昨日
                    } else {
                        tooOld = true // lookback+1 日目以降も missed = グレース超過
                        break loop
                    }
                default:
                    break loop // .future/.todayPending は後方走査では現れない想定。安全側で停止。
                }
            }
            let revivable = foundPrior && !tooOld && !missedOffsets.isEmpty
            let need = missedOffsets.count
            return Result(
                revivable: revivable,
                missedOffsets: revivable ? missedOffsets : [],
                freezesNeeded: revivable ? need : 0,
                hasEnough: revivable && remainingFreezes >= need
            )
        }
    }

    /// records から status 列(昨日→過去)を作り `Decision` に委譲する本番入口。
    /// anchor 検出のため lookback+1 日ぶん作る(missed 収集は Decision 側で lookback に制限)。
    static func evaluate(
        records: [WorkoutRecord],
        today: Date,
        rescuedDates: Set<Date>,
        remainingFreezes: Int,
        lookback: Int = 4,
        calendar: Calendar = .mondayFirst
    ) -> Result {
        let todayStart = calendar.startOfDay(for: today)
        var statuses: [DailyStatus] = []
        for offset in 1...(lookback + 1) {
            guard let day = calendar.date(byAdding: .day, value: -offset, to: todayStart) else { break }
            let restDays = RestDayResolver.restDaySet(for: day, records: records, today: todayStart, calendar: calendar)
            let s = AchievementEvaluator.dailyStatus(
                for: day, records: records, restDays: restDays,
                rescuedDates: rescuedDates, today: todayStart, calendar: calendar)
            statuses.append(s)
        }
        return Decision.evaluate(statuses: statuses, remainingFreezes: remainingFreezes, lookback: lookback)
    }

    /// offset 群(1=昨日…)を実際の Date(startOfDay)に変換。
    static func missedDates(forOffsets offsets: [Int], today: Date, calendar: Calendar = .mondayFirst) -> [Date] {
        let todayStart = calendar.startOfDay(for: today)
        return offsets.compactMap { calendar.date(byAdding: .day, value: -$0, to: todayStart) }
    }
}

File: app-android/app/src/main/java/com/goexercise/app/domain/StreakFreezeWindow.kt
package com.goexercise.app.domain

import java.time.LocalDate

/**
 * 連続記録のフリーズ(保険チケット)復活判定。iOS `StreakFreezeWindow` の純ロジック移植。
 *
 * 「直近に途切れた連続記録を、フリーズ(課金チケット)で復活できるか」を評価する。
 * 純判定層(`Decision`)と、記録から状態列を組み立てる層を分ける(前者は完全に純粋でテスト容易)。
 */
object StreakFreezeWindow {

    /**
     * 復活判定の結果。
     * @param revivable 復活可能(過去に達成があり、間に Missed が挟まる)か。
     * @param missedOffsets 復活に必要な Missed の日オフセット(1=昨日, 2=一昨日…)。
     * @param freezesNeeded 必要なフリーズ枚数(= missedOffsets.size)。
     * @param hasEnough 手持ちフリーズが必要数以上か。
     */
    data class Result(
        val revivable: Boolean,
        val missedOffsets: List<Int>,
        val freezesNeeded: Int,
        val hasEnough: Boolean,
    )

    /**
     * 純判定層。statuses[0]=offset1(昨日), [1]=offset2(一昨日)… の順。
     * statuses は offset 1..(lookback+1) を想定(アンカー検出用に 1 件余分に渡す)。
     * offset = index+1 として走査する。
     *
     * - Achieved / TodayAchieved → 過去の達成(アンカー)を発見し打ち切る(foundPrior=true)。
     * - Rest → スキップ(フリーズ不要、連続も切れない)。
     * - Missed → offset <= lookback なら復活対象として収集。それを超えていれば猶予枠超過で
     *   tooOld=true として打ち切る(lookback ちょうどの途切れは復活可能、それより古いと不可)。
     * - その他(Future / TodayPending)→ 走査打ち切り。
     *
     * revivable = アンカーあり && !tooOld && Missed が 1 件以上。
     */
    object Decision {
        fun evaluate(
            statuses: List<DailyStatus>,
            remainingFreezes: Int,
            lookback: Int,
        ): Result {
            val missedOffsets = mutableListOf<Int>()
            var foundPrior = false
            var tooOld = false

            run {
                statuses.forEachIndexed { idx, status ->
                    val offset = idx + 1
                    when (status) {
                        DailyStatus.Achieved, DailyStatus.TodayAchieved, DailyStatus.Rescued -> {
                            // rescued(過去のフリーズ救済日)も達成と同じ「連続の頭」(表示分離用)。iOS パリティ。
                            foundPrior = true
                            return@run
                        }
                        DailyStatus.Rest -> {
                            // skip — フリーズ不要、連続も切れない
                        }
                        DailyStatus.Missed -> {
                            if (offset <= lookback) {
                                missedOffsets.add(offset)
                            } else {
                                // 猶予枠を超える途切れ → 復活不可
                                tooOld = true
                                return@run
                            }
                        }
                        else -> {
                            // Future / TodayPending → 打ち切り
                            return@run
                        }
                    }
                }
            }

            val revivable = foundPrior && !tooOld && missedOffsets.isNotEmpty()
            if (!revivable) {
                return Result(
                    revivable = false,
                    missedOffsets = emptyList(),
                    freezesNeeded = 0,
                    hasEnough = false,
                )
            }

            val needed = missedOffsets.size
            return Result(
                revivable = true,
                missedOffsets = missedOffsets.toList(),
                freezesNeeded = needed,
                hasEnough = remainingFreezes >= needed,
            )
        }
    }

    /**
     * 記録から状態列を組み立てて判定を委譲する。
     * rest 日は lookback 枠を消費しない(連続を切らない)ため、anchor(連続の頭=achieved/rescued)に
     * 到達するまで rest を読み飛ばして走査を**動的に延長**する。固定 lookback+1 件だと、
     * missed と anchor の間に自動休養(週最大2日)が挟まったとき anchor が窓の外へ押し出され、
     * 復活可能なのに復活ポップが発火しない(iOS が修正済みの不具合の Android 未移植分)。
     * 打ち切り条件: anchor 到達 / lookback 超の missed 到達 / 安全上限(lookback + 1週間)。
     */
    fun evaluate(
        records: List<WorkoutRecord>,
        today: LocalDate,
        rescuedDates: Set<LocalDate>,
        remainingFreezes: Int,
        lookback: Int = 4,
    ): Result {
        val statuses = mutableListOf<DailyStatus>()
        // 自動休養は最大週2日。grace(lookback)+ 連続した休養の最大幅を吸収できる安全上限。
        val hardCap = lookback + 7
        for (offset in 1..hardCap) {
            val date = today.minusDays(offset.toLong())
            val restDays = RestDayResolver.restDaySet(date, records, today)
            val s = AchievementEvaluator.dailyStatus(
                date = date,
                records = records,
                restDays = restDays,
                rescuedDates = rescuedDates,
                today = today,
            )
            statuses.add(s)
            when (s) {
                DailyStatus.Achieved, DailyStatus.TodayAchieved, DailyStatus.Rescued -> break // anchor 到達
                DailyStatus.Rest -> Unit // 休養は枠を消費せず連続も切らない → 延長
                DailyStatus.Missed -> if (offset > lookback) break // グレース超 → 復活不可確定
                else -> break // Future / TodayPending は後方に現れない想定、安全側で停止
            }
        }
        return Decision.evaluate(statuses, remainingFreezes, lookback)
    }

    /** Missed オフセット群を実日付へ変換。 */
    fun missedDatesForOffsets(offsets: List<Int>, today: LocalDate): List<LocalDate> =
        offsets.map { today.minusDays(it.toLong()) }
}
